### PR TITLE
LIF ProcessModels: Floating and fixed point: PR attempt #2

### PR DESCRIPTION
**Pull request by srrisbud**
There are merge conflicts. Trying to resolve these gracefully, without using the nuclear option of deleting and recreating the entire fork.

### CODE REVIEW

**Comment by srrisbud** [comment]
> There are merge conflicts. Trying to resolve these gracefully, without using the nuclear option of deleting and recreating the entire fork.

Seems like I was successful in fixing the merge issues. There are no merge conflicts anymore.

File: src/lava/proc/lif/models.py [hunk 1]
```diff
@@ -7,29 +7,131 @@
...
+from lava.magma.core.decorator import implements, requires, tag
 from lava.magma.core.model.py.model import PyLoihiProcessModel
 from lava.proc.lif.process import LIF
 
 
 @implements(proc=LIF, protocol=LoihiProtocol)
 @requires(CPU)
-class PyLifModel(PyLoihiProcessModel):
+@tag('floating_pt')
+class PyLifModelFloat(PyLoihiProcessModel):
```
Comment (mathisrichter) [inline src/lava/proc/lif/models.py:18]: Could you write a docstring here to explain what the point of this ProcessModel is (default Process Model based on Python and using floating point numbers)?

File: src/lava/proc/lif/models.py [hunk 2]
```diff
@@ -7,29 +7,131 @@
...
+            decayed_volt), 12)
+        decayed_volt = np.int32(decayed_volt)
+        updated_volt = decayed_volt + self.u + self.effective_bias
+        self.v[:] = np.clip(updated_volt, neg_voltage_limit, pos_voltage_limit)
+
+        # Spike when exceeds threshold
+        # ----------------------------
+        s_out = self.v >= self.effective_vth
+        # Reset voltage of spiked neurons to 0
+        self.v[s_out] = 0
```
Comment (mathisrichter) [inline src/lava/proc/lif/models.py:136]: Can these magic bit numbers come from descriptive constants instead?

File: tests/lava/proc/lif/test_process.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+import unittest
+import numpy as np
+from lava.proc.lif.process import LIF
+
+
+class TestLIFProcess(unittest.TestCase):
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_process.py:9]: This tests only the initialization of the Process. What about the Process Models and their behavior?

**Review by mathisrichter — CHANGES_REQUESTED** [review]
Looks good in general.
The most important request from my side are unit tests for the Process Models and their dynamics.

I didn't check the changes to the tutorials.

File: src/lava/proc/lif/models.py (hunk 1 above)
Comment (srrisbud) [reply src/lava/proc/lif/models.py:18]: Done.

File: src/lava/proc/lif/models.py (hunk 2 above)
Comment (srrisbud) [reply src/lava/proc/lif/models.py:136]: Done. Currently defined all magic numbers as instance variables of a ProcessModel instance. We can think of reusing enums from NxCompiler, when that is ready.

File: tests/lava/proc/lif/test_process.py (hunk 3 above)
Comment (srrisbud) [reply tests/lava/proc/lif/test_process.py:9]: They are in a different file: tests/lava/proc/lif/test_models.py

**Review by bamsumit — APPROVED** [review]
Looks good to me. I didn't go through the notebooks. I guess they work ;)

**Review by srrisbud — COMMENTED** [review]
Went over everything once

**Comment by mathisrichter** [comment]
Nice! But no unit tests for the two Process Models?

**Review by mathisrichter — APPROVED** [review]

**Comment by mathisrichter** [comment]
> Nice! But no unit tests for the two Process Models?

Never mind, I overlooked all of that. Sorry!

File: src/lava/proc/lif/models.py (hunk 2 above)
Comment (awintel) [reply src/lava/proc/lif/models.py]: Perhaps start pulling them out of the LIF class at least into a local ENUM within this module so we find their usage later more easily once we use a more systematic constant database.

**Comment by bamsumit** [comment]
@srrisbud the new lif unittests are really polluting the console with deprication warnings. Apparently they don't show up in `pyb -E unit` but try `python -m unittest discover test/` and it shows up and there are loads of them.

```bash
/home/sshresth/lava-nc/lava/src/lava/proc/lif/models.py:24: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  a_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, np.float)
/home/sshresth/lava-nc/lava/src/lava/proc/lif/models.py:26: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  u: np.ndarray = LavaPyType(np.ndarray, np.float)
```

**Pull request merged**